src/interpolations.jl

```diff
@@ -1264,6 +1250,429 @@ function shape_value(ip::BubbleEnrichedLagrange{RefTriangle, 1}, ξ::Vec{2}, i::
+
+for (name,  default_coords) in (
+    (:ArbitraryOrderLagrange,               :both),
+    (:ArbitraryOrderDiscontinuousLagrange,  :neither)
+    )
+    @eval begin
+        function $(name){RefLine, order}(points::Vector{Float64} = GaussQuadrature.legendre(order+1, GaussQuadrature.$(default_coords))[1]) where order
+            product_of = nothing
+            ref_coord = Array{Vec{1,Float64},1}(undef,order+1)
+            for i in 1:order+1
+                ref_coord[i] = Vec(points[i])
+            end
+            perm = getPermLagrangeLine(order)
+            inv_perm = sortperm(perm)
+            $(name){RefLine, order, typeof(product_of), typeof(ref_coord), typeof(perm)}(product_of, ref_coord, perm, inv_perm)
+        end
+        getnbasefunctions(::$(name){RefLine,order}) where order = order + 1
+    end
+end
+
+for (name,  facefunc) in (
+    (:ArbitraryOrderLagrange,               :facedof_indices          ),
+    (:ArbitraryOrderDiscontinuousLagrange,  :dirichlet_facedof_indices)
+    )
+    @eval begin
+        $(facefunc)(::$(name){RefLine,order}) where order = ((1,), (2,))
+        function shape_value(ip::$(name){RefLine, order}, ξ::Vec{1}, j::Int) where order
+            j > getnbasefunctions(ip) && throw(ArgumentError("no shape function $j for interpolation $ip"))
+            ξ_x = ξ[1]
+            result = 1.0
+            j = ip.perm[j]
+            coeff = ip.reference_coordinates
+            for k in 1:order+1
+                k == j && continue
+                result *= (ξ_x - coeff[k][1])/(coeff[j][1]-coeff[k][1])
+            end
+            return result
+        end
+    end
+end
+
+celldof_interior_indices(::ArbitraryOrderLagrange{RefLine,order}) where order = SVector{order-1}(i+2 for i in 1:order - 1)        
+
+##############################
+# Lagrange RefQuadrilateral  #
+##############################
+
+# Permutation to switch numbering to Ferrite ordering
+function getPermLagrangeQ(order)
+    result = Array{Int,1}(undef, (order+1)^2)
+    result[1:4] .= (1, order + 1, (order + 1)^2, order * (order + 1) + 1)
+    idx = 5
+    for i in 2:order
+        result[idx] = i
+        idx += 1
+    end
+    for i in 2:order
+        result[idx] = (order+1)*i
+        idx += 1
+    end
+    for i in 1:order-1
+        result[idx] = (order+1)^2-i
+        idx += 1
+    end
+    for i in order-1:-1:1
+        result[idx] = (order+1)*i+1
+        idx += 1
+    end
+    for j in 1:(order-1), i in 1:(order-1)
+        result[idx] = i+1+(order+1)*j
+        idx += 1
+    end
+    return result
+end
+
+
+for (name,  default_coords) in (
+    (:ArbitraryOrderLagrange,               :both),
+    (:ArbitraryOrderDiscontinuousLagrange,  :neither)
+    )
+    @eval begin
+        function $(name){RefQuadrilateral, order}(points::Vector{Float64} = GaussQuadrature.legendre(order+1, GaussQuadrature.$(default_coords))[1]) where order
```

Comment:
Can we add doc strings for these, so users know about the functionality?

Also, how can users discover this feature? Maybe we should add some hint in the Lagrange doc string and a tip one of the examples? Not sure if it is worth to copy paste the full heat example here. In the future we can think about adding a Lagrangian formulation of some advection problem.